File: assets/zig-kernel-feasibility/interop.c
#include "abi.h"
uint64_t c_add(uint64_t a, uint64_t b) { return a + b; }
struct pair c_apply(struct pair p, callback cb) {
    p.value = cb(p.value);
    p.tag += 1;
    return p;
}
#ifdef HOST_TEST
#include <stdio.h>
int main(void) {
    struct pair out = zig_roundtrip((struct pair){7, 40});
    if (out.tag != 8 || out.value != 42 || zig_callback(9) != 11) return 1;
    puts("PASS: C -> Zig -> C -> Zig callback; by-value struct and layout");
    return 0;
}
#endif

File: assets/zig-kernel-feasibility/interop.zig
const std = @import("std");
pub const panic = std.debug.FullPanic(struct {
    fn stop(_: []const u8, _: ?usize) noreturn { @trap(); }
}.stop);
const Pair = extern struct { tag: u32, value: u64 };
comptime {
    if (@sizeOf(Pair) != 16 or @alignOf(Pair) != 8 or @offsetOf(Pair, "value") != 8)
        @compileError("pair layout mismatch");
}
extern fn c_add(a: u64, b: u64) u64;
extern fn c_apply(p: Pair, cb: *const fn (u64) callconv(.c) u64) Pair;
export fn zig_callback(n: u64) u64 { return c_add(n, 2); }
export fn zig_roundtrip(p: Pair) Pair { return c_apply(p, &zig_callback); }
